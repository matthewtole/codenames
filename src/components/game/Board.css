.BoardWrapper {
  flex: 1 1;
}

.Board {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  padding: 5px;
}

.BoardRow {
  display: flex;
  flex-direction: row;
  height: 20%;
}

@media only screen and (max-width: 1024px) {
  .Board {
    padding: 3px;
  }
}
