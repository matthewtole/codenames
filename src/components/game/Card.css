.Card {
  width: 20%;
  font-size: 1.75rem;
  /* margin: 1px; */
  user-select: none;
  position: relative;
}

.Card > div {
  background-color: #eee;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  top: 0;
  border-radius: 4px;
  transition: margin 0.1s ease-out;
  margin: 0;
  border: 2px solid #fff;
}

.Card--revealed {
  text-decoration: line-through;
  opacity: 0.2;
}

.Card--controller[data-role='BLUE_SPY'] > div {
  background-color: #bbdefb;
}

.Card--controller[data-role='RED_SPY'] > div {
  background-color: #ef9a9a;
}

.Card--controller[data-role='BYSTANDER'] > div {
  background-color: #ff9100;
}

.Card--controller[data-role='ASSASSIN'] > div {
  background-color: #999;
}

.Card[data-role='BLUE_SPY'].Card--revealed > div {
  background-color: #2196f3;
  color: #bbdefb;
}

.Card[data-role='RED_SPY'].Card--revealed > div {
  background-color: #d32f2f;
  color: #ef9a9a;
}

.Card[data-role='BYSTANDER'].Card--revealed > div {
  background-color: #ff9100;
}

.Card[data-role='ASSASSIN'].Card--revealed > div {
  background-color: #222;
}

.Card--highlighted > div {
  /* border: 5px solid #000; */
  /* box-shadow: 0 0 20px rgba(0, 0, 0, 0.5); */
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  margin: -7px;
  z-index: 200;
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;
  }
}
