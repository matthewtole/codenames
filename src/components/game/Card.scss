@import '../../sass/_variables';

.Card {
  width: 20%;
  font-size: 1.75rem;
  user-select: none;
  position: relative;

  & > div {
    background-color: #eee;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    border-radius: 4px;
    margin: 5px;
    border: 2px solid #ddd;

    box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
    transition: all 0.3s cubic-bezier(0.25, 0.8, 0.25, 1);
  }

  &--revealed > div {
    border: 0;
  }

  &--highlighted > div {
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
    // margin: -7px;
    z-index: 10;
  }

  &--revealed {
    text-decoration: line-through;
    opacity: 0.2;
  }

  &[data-role='BLUE_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $blue-spy--light;
        border-color: $blue-spy--dark;
      }

      &.Card--controller.Card--highlighted > div {
        background-color: $blue-spy--dark;
        color: #fff;
      }
    }

    &.Card--revealed {
      > div {
        background-color: $blue-spy--dark;
        color: $blue-spy--light;
      }
    }
  }

  &[data-role='RED_SPY'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $red-spy--light;
        border-color: $red-spy--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $red-spy--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $red-spy--dark;
        color: $red-spy--light;
      }
    }
  }

  &[data-role='BYSTANDER'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $bystander--light;
        border-color: $bystander--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $bystander--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $bystander--dark;
      }
    }
  }

  &[data-role='ASSASSIN'] {
    &.Card--controller,
    &.Card--game-over {
      > div {
        background-color: $assassin--light;
        border-color: $assassin--dark;
      }
    }

    &.Card--controller.Card--highlighted > div {
      background-color: $assassin--dark;
      color: #fff;
    }

    &.Card--revealed {
      > div {
        background-color: $assassin--dark;
      }
    }
  }
}

@media only screen and (max-width: 1024px) {
  .Card {
    font-size: 1.2rem;

    & > div {
      margin: 3px;
    }
  }
}
