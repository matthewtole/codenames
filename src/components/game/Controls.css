.GameControls {
  background-color: #ddd;
  height: 64px;
  display: flex;
  flex-shrink: 0;
  justify-content: center;
}

.GameControls_end-turn {
  height: 100%;
  flex: 2 1;
  background-color: #333;
  color: #fff;
  border: 0;
  font-size: 22px;
  letter-spacing: 2px;
  font-weight: 900;
  text-transform: uppercase;
}

.GameControls_end-turn:hover {
  background-color: #888;
}

.GameControls_new-game {
  height: 100%;
  flex: 1 1;
  background-color: #fff;
  color: #333;
  border: 0;
  font-size: 22px;
  letter-spacing: 2px;
  font-weight: 900;
  text-transform: uppercase;
}

.GameControls_new-game:hover {
  background-color: #ccc;
}

.GameControls_menu {
  height: 100%;
  flex: 0 0;
  background-color: #fff;
  color: #333;
  border: 0;
  font-size: 22px;
  letter-spacing: 2px;
  font-weight: 900;
  text-transform: uppercase;
}

.GameControls_menu:hover {
  background-color: #ccc;
}

@media only screen and (max-width: 1024px) {
  .GameControls_end-turn {
    font-size: 16px;
  }
  .GameControls_new-game {
    font-size: 16px;
  }
}
