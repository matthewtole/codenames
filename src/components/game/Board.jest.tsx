import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { BoardProps, BoardMode, Board } from './Board';
import { OriginalCards1 } from '../../fixtures/cards';

describe('Board', () => {
  const defaultProps: BoardProps = {
    cards: OriginalCards1,
    width: 5,
    height: 5,
    revealedCards: [],
    mode: BoardMode.CONTROLLER,
    isGameOver: false,
    onRevealCard: jest.fn(),
    onHighlightCard: jest.fn(),
  };

  it('renders correctly', () => {
    const tree = renderer.create(<Board {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
