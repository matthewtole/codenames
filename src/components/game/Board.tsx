import * as React from 'react';
import './Board.css';
import { Card as CardComponent } from './Card';
import {
  Card,
  Coordinate,
  Role,
  CoordinateValue,
} from '../../redux/game/types';

export enum BoardMode {
  VIEWER = 'VIEWER',
  CONTROLLER = 'CONTROLLER',
}

export interface BoardProps {
  width: number;
  height: number;
  cards: Card[];
  revealedCards: number[];
  mode: BoardMode;
  highlighted?: Coordinate;
  highlightedRow?: number;
  isGameOver: boolean;
  onRevealCard: (card: Coordinate) => void;
  onHighlightCard: (card?: Coordinate) => void;
}

export class Board extends React.PureComponent<BoardProps, {}> {
  constructor(props: BoardProps) {
    super(props);
    this.state = {};
  }

  render() {
    return (
      <div className="BoardWrapper" onClick={() => this.highlightCard()}>
        <div className="Board">
          {this.renderRowHighlight()}
          {Array(this.props.height)
            .fill(0)
            .map((_: number, row: CoordinateValue) => this.renderRow(row))}
        </div>
      </div>
    );
  }

  private renderRowHighlight() {
    if (this.props.highlightedRow === undefined) {
      return null;
    }
    const row = this.props.highlightedRow;
    return (
      <div className="Board__row-highlight-wrapper">
        <div
          className="Board__row-highlight"
          style={{ top: 0, height: `${row * 20}%` }}
        />
        <div
          className="Board__row-highlight"
          style={{
            top: `${(row + 1) * 20}%`,
            height: `${(5 - row + 1) * 20}%`,
          }}
        />
      </div>
    );
  }

  private renderRow = (row: CoordinateValue) => {
    return (
      <div className="BoardRow" key={`row_${row}`}>
        {Array(this.props.width)
          .fill(0)
          .map((__: number, col: CoordinateValue) => this.renderCard(row, col))}
      </div>
    );
  }

  private renderCard = (row: CoordinateValue, col: CoordinateValue) => {
    const index = this.coordinateToIndex({ row, col });
    return (
      <CardComponent
        key={`card_${row}_${col}`}
        word={this.getWord(index)}
        role={this.getRole(index)}
        revealed={this.isRevealed(index)}
        highlighted={this.isHighlighted({ row, col })}
        onHighlight={() => {
          this.highlightCard({ row, col });
        }}
        onReveal={() => {
          this.revealCard({ row, col });
        }}
        boardMode={this.props.mode}
        isGameOver={this.props.isGameOver}
      />
    );
  }

  private getWord = (index: number): string => {
    return this.props.cards[index].word;
  }

  private getRole = (index: number): Role => {
    return this.props.cards[index].role;
  }

  private isRevealed = (index: number): boolean => {
    return this.props.revealedCards.indexOf(index) >= 0;
  }

  private isHighlighted = (card: Coordinate): boolean => {
    if (this.props.highlighted === undefined) {
      return false;
    }
    return (
      this.props.highlighted.row === card.row &&
      this.props.highlighted.col === card.col
    );
  }

  private highlightCard = (card?: Coordinate) => {
    this.props.onHighlightCard(card);
  }

  private revealCard = (card: Coordinate) => {
    this.props.onRevealCard(card);
  }

  private coordinateToIndex(coordinate: Coordinate): number {
    return coordinate.row * this.props.width + coordinate.col;
  }
}
