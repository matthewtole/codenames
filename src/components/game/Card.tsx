import * as React from 'react';
import './Card.css';
import { Role } from '../../lib/types';
import * as cx from 'classnames';
import { BoardMode } from './Board';

export interface CardProps {
  word: string;
  role: Role;
  revealed: boolean;
  highlighted: boolean;
  boardMode: BoardMode;
  onReveal: () => void;
  onHighlight: () => void;
}

export class Card extends React.PureComponent<CardProps, object> {
  render() {
    return (
      <div
        className={cx({
          Card: true,
          'Card--highlighted': this.props.highlighted,
          'Card--revealed': this.props.revealed,
          [`Card--${this.props.boardMode}`]: true,
        })}
        data-role={this.props.role}
        onClick={this.handleClick}
      >
        <div>{this.props.word}</div>
      </div>
    );
  }

  handleClick = (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    if (this.props.revealed) {
      return;
    }
    if (this.props.highlighted) {
      this.props.onReveal();
    } else {
      this.props.onHighlight();
    }
  }
}
