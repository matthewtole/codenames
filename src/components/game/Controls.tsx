import * as React from 'react';
import { Team } from '../../redux/game/types';
import './Controls.css';

interface Props {
  winner?: Team;
  onEndTurn: () => void;
  onNewGame: () => void;
  onMenuOpen: () => void;
  showMenu?: boolean;
}

export class Controls extends React.Component<Props, {}> {
  render() {
    const { winner, onEndTurn, onNewGame, onMenuOpen, showMenu } = this.props;
    const controls = [];
    if (!winner) {
      controls.push(
        <button
          key="end-turn"
          className="GameControls_end-turn"
          onClick={() => onEndTurn()}
        >
          End Turn
        </button>
      );
    }
    controls.push(
      <button
        key="new-game"
        className="GameControls_new-game"
        onClick={() => onNewGame()}
      >
        Start New Game
      </button>
    );
    if (showMenu) {
      controls.push(
        <button
          className="GameControls_menu"
          onClick={() => onMenuOpen()}
          key="menu"
        >
          <i className="fas fa-bars fa-fw fa-lg" />
        </button>
      );
    }
    return <div className="GameControls">{controls}</div>;
  }
}
