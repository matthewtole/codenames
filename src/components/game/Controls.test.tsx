import * as React from 'react';
import { Controls } from './Controls';
import * as renderer from 'react-test-renderer';
import { shallow, ShallowWrapper } from 'enzyme';
import { Team } from '../../redux/game/types';

describe('Controls', () => {
  const onEndTurn = jest.fn();
  const onNewGame = jest.fn();
  const onMenuOpen = jest.fn();

  const defaultProps = {
    onEndTurn,
    onNewGame,
    onMenuOpen,
  };

  beforeEach(() => {
    onEndTurn.mockReset();
    onNewGame.mockReset();
  });

  function getEndTurnButton(controls: ShallowWrapper) {
    return controls.find('button').filterWhere(n => n.text() === 'End Turn');
  }

  function getNewGameButton(controls: ShallowWrapper) {
    return controls
      .find('button')
      .filterWhere(n => n.text() === 'Start New Game');
  }

  it('renders correctly', () => {
    const controls = renderer.create(<Controls {...defaultProps} />).toJSON();
    expect(controls).toMatchSnapshot();
  });

  it('should render the End Turn button when there is no winner', () => {
    const controls = shallow(<Controls {...defaultProps} />);
    expect(controls.text()).toContain('End Turn');
  });

  it('should not render the End Turn button when there is a winner', () => {
    const controls = shallow(<Controls {...defaultProps} winner={Team.BLUE} />);
    expect(controls.text()).not.toContain('End Turn');
  });

  it('should render the New Game button when there is no winner', () => {
    const controls = shallow(<Controls {...defaultProps} />);
    expect(controls.text()).toContain('New Game');
  });

  it('should render the New Game button when there is a winner', () => {
    const controls = shallow(<Controls {...defaultProps} winner={Team.BLUE} />);
    expect(controls.text()).toContain('New Game');
  });

  it('should call onEndTurn when clicking the End Turn button', () => {
    const controls = shallow(<Controls {...defaultProps} />);
    getEndTurnButton(controls).simulate('click');
    expect(onEndTurn).toHaveBeenCalled();
  });

  it('should call onNewGame when clicking the Start New Game button', () => {
    const controls = shallow(<Controls {...defaultProps} winner={Team.BLUE} />);
    getNewGameButton(controls).simulate('click');
    expect(onNewGame).toHaveBeenCalled();
  });
});
