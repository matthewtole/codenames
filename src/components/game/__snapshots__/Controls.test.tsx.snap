// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls renders correctly 1`] = `
<div
  className="GameControls"
>
  <button
    className="GameControls_end-turn"
    onClick={[Function]}
  >
    End Turn
  </button>
  <button
    className="GameControls_new-game"
    onClick={[Function]}
  >
    Start New Game
  </button>
  <button
    className="GameControls_menu"
    onClick={[Function]}
  >
    <i
      className="fas fa-bars fa-fw fa-lg"
    />
  </button>
</div>
`;
